Bug fix: Emit code for C++ inline function called from another inline function

martinboehme opened the pull request:
Adds a test that fails without the fix.

The fix consists simply of removing a return statement; as a result, we now always emit code for functions referenced from the function we are currently emitting.

Previously, this was not done for externally visible functions. This goes all the way back to when `GenClangDecl.cpp` was originally introduced in https://github.com/apple/swift/commit/ee22004b84dd72fd9d99dbb11837f5dfd377d08a.

I speculate that the reason for doing this was that in C and Objective-C, we can assume that an externally visible function will be provided by some `.o` file we're linking against. This is also true for `inline` functions in C99, where exactly one translation unit needs to declare the function `extern inline` to cause it to get emitted to the `.o` file.

C++ inline functions follow a different model, however, so we need to make sure we emit code for them.

martinboehme committed:
Bug fix: Emit code for C++ inline function called from another inline function.

Adds a test that fails without the fix.

The fix consists simply of removing a return statement; as a result, we
now always emit code for functions referenced from the function we are
currently emitting.

Previously, this was not done for externally visible functions. This
goes all the way back to when `GenClangDecl.cpp` was originally introduced here:

https://github.com/apple/swift/commit/ee22004b84dd72fd9d99dbb11837f5dfd377d08a

I speculate that the reason for doing this was that in C and Objective-C, we
can assume that an externally visible function will be provided by some
`.o` file we're linking against (though I'm not sure if this is in fact
true for C99 inline functions). At any rate, this assumption is no longer true
for C++ inline functions.

MForster commented on test/Interop/Cxx/function/Inputs/module.modulemap:
```diff
@@ -0,0 +1,4 @@
+module EmitCalledFunction {
+  header "EmitCalledFunction.h"
```
(nit) Should we keep to a consistent naming convention for files (lowercase with dashes)?

martinboehme replied:
Good point -- done.

gribozavr commented:
@swift-ci Please smoke test

gribozavr commented on test/Interop/Cxx/function/emit-called-function-ir.swift:
```diff
@@ -0,0 +1,14 @@
+// This is a reproducer for a bug where Swift was not emitting code for a
```
Could you rephrase this as a test for emitting inline functions that are used in general? The "regression test" wording is useful for obscure combinations, but this seems like a pretty basic case that we should be testing regardless.

martinboehme replied:
Good point, done.

gribozavr commented on test/Interop/Cxx/function/emit-called-function-ir.swift:
```diff
@@ -0,0 +1,14 @@
+// This is a reproducer for a bug where Swift was not emitting code for a
```
s/-ir/-irgen/ in the file name.

martinboehme replied:
Done.

martinboehme committed:
Rename EmitCalledFunction.h to emit-called-function.h.

gribozavr commented on test/Interop/Cxx/function/Inputs/EmitCalledFunction.h:
```diff
@@ -0,0 +1,7 @@
+
+inline int foo() {
+  return bar();
+}
```
Please use more descriptive names than foo and bar.

martinboehme replied:
Done.

While I was doing this, it occurred to me that I should also add a function that isn't called from Swift at all (whether directly or indirectly) and test that it isn't emitted.

martinboehme committed:
Respond to review comments by @gribozavr.

gribozavr commented on test/Interop/Cxx/function/emit-called-function-irgen.swift:
```diff
@@ -0,0 +1,20 @@
+// Test that Swift emits code for a C++ inline function that is called from
```
Please add "inline" to the names of the tests.

martinboehme replied:
Done.

martinboehme commented:
@rjmccall I'd like your feedback on whether this fix looks right.

gribozavr commented on test/Interop/Cxx/function/emit-called-function-irgen.swift:
```diff
@@ -0,0 +1,20 @@
+
+import EmitCalledFunction
+
+// Unfortunately, we have to repeat the CHECK-NOT to express that notCalled()
```
Please use `--implicit-check-not`.

martinboehme replied:
Done.

gribozavr commented on test/Interop/Cxx/function/emit-called-function-irgen.swift:
```diff
@@ -0,0 +1,20 @@
+// Test that Swift emits code for a C++ inline function that is called from
```
"Test that we emit LLVM IR for C++ inline free functions that are called directly, and also for inline functions that are called transitively. Test that we don't emit LLVM IR for inline functions that are not called."

martinboehme replied:
Thanks, this is much clearer. Done, with some additional wordsmithing that I hope you're OK with.

gribozavr commented on test/Interop/Cxx/function/emit-called-function-irgen.swift:
```diff
@@ -0,0 +1,20 @@
+// indirectly from Swift.
+
+// RUN: %empty-directory(%t)
+// RUN: %target-swift-frontend %s -I %S/Inputs -enable-cxx-interop -emit-ir -o - | %FileCheck %s
```
WDYT about running this test in both C and C++ modes?

martinboehme replied:
Good point. Done.

I'm using `static inline` in C, as `extern inline` follows a [different model](https://gustedt.wordpress.com/2010/11/29/myth-and-reality-about-inline-in-c99/), where there's always one translation unit that explicitly causes an out-of-line instance to be emitted (so it doesn't have the issue we're looking at here).

`static inline` C functions aren't affected by the bug I'm fixing here as they don't have external linkage, so they don't fall into the case that does the early-out. It's still good to have this test in place though.

gribozavr commented on test/Interop/Cxx/function/Inputs/emit-called-function.h:
```diff
@@ -0,0 +1,11 @@
+  return 42;
+}
+
+inline int calledIndirectly() {
```
s/indirectly/transitively/

"Indirectly" means "through a pointer".

martinboehme replied:
Good point. Done.

martinboehme committed:
More reponses to review comments by @gribozavr.

One larger change is that we're now also running this test in C as well
as C++.

rjmccall commented on lib/IRGen/GenClangDecl.cpp:
```diff
@@ -41,7 +41,6 @@ void IRGenModule::emitClangDecl(const clang::Decl *decl) {
     ClangCodeGen->HandleTopLevelDecl(
                           clang::DeclGroupRef(const_cast<clang::Decl*>(decl)));
-    return;
   }
```
This looks correct, but can we preserve the fast path where the declaration does not have a definition?

martinboehme replied:
I assume you mean the fast path where the declaration isn't a `ValueDecl` (which I've done)?

There wasn't a fast path before for the case where the declaration doesn't have a definition (and I believe this isn't possible to check in the general case of a `ValueDecl`?).

rjmccall replied:
Well, the fast-path was buggy, we've established that. :)  I think the point of the `isExternallyVisible()` check was supposed to be to look for a function declaration like `void foo();` that declares something with external linkage, but of course that's not the right way to check for that.

Currently the rest of the function only does special things for `FunctionDecl`s, although really it needs to do them for `VarDecl`s as well.  I think having a helper function that tries to find a body `Stmt*` for a `ValueDecl` wouldn't be remiss.

martinboehme replied:
> Well, the fast-path was buggy, we've established that. :) I think the point of the `isExternallyVisible()` check was supposed to be to look for a function declaration like `void foo();` that declares something with external linkage, but of course that's not the right way to check for that.

Ah -- that makes sense.

> Currently the rest of the function only does special things for `FunctionDecl`s, although really it needs to do them for `VarDecl`s as well.

Oooooh, right! I've verified this case was not being handled correctly either. See the extra test case I've added -- `calledTransitivelyFromVarInit()` was not being emitted before I made the code changes in my latest commit.

> I think having a helper function that tries to find a body `Stmt*` for a `ValueDecl` wouldn't be remiss.

A `VarDecl` doesn't have a `Stmt *` though...? But I see what you mean, and I've introduced a helper function that returns a redeclaration of the decl containing "executable code" (body for a function, initializer for a variable), if one exists. This decl contains the body (in the case of a function) or initializer (in the case of a variable), so it's the right thing to traverse to find DeclRefs of things we should emit.

Not super-pleased with the naming of the helper function BTW (`getDeclWithExecutableCode()`) -- let me know if you have a better suggestion.

rjmccall replied:
A quirk of Clang's AST is that `Expr` is a subclass of `Stmt`, so `VarDecl`s *do* actually have a `Stmt*`, but this is a fine approach, too.

martinboehme replied:
Oh right, I had forgotten about that -- makes sense, given that an expression is a valid statement in C++.

martinboehme committed:
Respond to review comment by @rjmccall.

Reinstate fast path where declaration is not a ValueDecl.

gribozavr commented on test/Interop/C/function/Inputs/emit-called-function.h:
```diff
@@ -0,0 +1,23 @@
+}
+
+INLINE int calledFromSwift() {
+  return calledTransitively();
```
Since there is non-trivial logic in `ClangDeclRefFinder` that distinguishes toplevel inline functions and other functions, WDYT about adding another test where we call an inline member function from a toplevel inline function?

martinboehme replied:
Done.

martinboehme committed:
Responses to review comments by @gribozavr.

martinboehme commented on test/Interop/C/function/Inputs/module.modulemap:
```diff
@@ -0,0 +1,4 @@
+module EmitCalledFunction {
+  header "emit-called-function.h"
```
Done.

martinboehme committed:
Also emit inline functions that are used in variable initializers.

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test macOS

gribozavr commented:
@swift-ci Please test Windows

gribozavr commented:
@swift-ci Please test

martinboehme committed:
Make tests pass under Darwin and Windows.

gribozavr commented:
@swift-ci Please test

martinboehme committed:
Fix "CXX-DAG" lines mistakenly renamed to "CXX-DOG" for testing.

martinboehme committed:
Merge branch 'master' into emit-called-func

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test macOS

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@swift-ci Please test Windows

gribozavr commented:
Didn't notice that we already have a recent passing Windows build: https://ci-external.swift.org/job/swift-PR-windows/2043/

Merging.

rjmccall reviewed (commented):
LGTM
